### PR TITLE
More tests

### PR DESCRIPTION
**Pull request by nerdinand**

### CODE REVIEW

File: spec/models/day_spec.rb [hunk 1]
```diff
@@ -1,43 +1,44 @@
...
+        Day.new(students: [student])
       end
 
       it 'returns true' do
         expect(subject.has_student?(student)).to be true
       end
     end
-
+# so we let things which we need but aren't the object of the test and use
+# subject when we refer to an object under test?
```
Comment (nerdinand) [inline spec/models/day_spec.rb:19]: Yes, that's exactly the idea. If you want to know the details, this is a fantastic answer to the difference between `subject` and `let`: https://stackoverflow.com/a/38459039/1453644

Actually, that answer prefers to avoid `subject` altogether and advertises for using `let` instead. This is a question of personal taste. 

I guess the original, intent of how `subject` would be used, is like this:
```
subject do
  Day.new(students: [student1])
end

it 'returns false' do
  expect(subject.has_student?(student2).to be false
end
```

Again, you're free to decide what's clearer for you. In terms of functionality, it shouldn't make a difference.

File: spec/models/day_spec.rb [hunk 2]
```diff
@@ -1,43 +1,44 @@
...
+        expect(day.has_student?(student2)).to be false
       end
     end
   end
 
   describe '#students_without_attendance' do
     context 'day has no attendances' do
-      subject(:day) do
-        described_class.new
+      subject(:day) do # is this equivalent to let(:day) { Day.new } ???
```
Comment (nerdinand) [inline spec/models/day_spec.rb:36]: Yes, they're almost equivalent. The only difference would be that if you use with `let` you cannot use `subject` in the `it` block (see example and stackoverflow link above).

File: spec/models/day_spec.rb [hunk 3]
```diff
@@ -47,9 +48,52 @@
...
+      it 'returns nil' do
+        expect(day.next).to eq(nil)
+      end
+    end
+  end
+
+  describe '#previous' do
+    context 'a day that has previous day' do
+      subject(:day) do
+        days(:day2) # find out what days does
```
Comment (nerdinand) [inline spec/models/day_spec.rb:81]: `days` is a method defined through the line `fixtures :days, :students, :attendances` at the beginning of the file. It lets you access fixtures with a specific name (here it's a days fixture with the name "day2"). Fixtures are just records stored in the test database (when you start the tests, they're loaded automatically from the `spec/fixtures/` directory).

File: spec/models/glider_spec.rb [hunk 4]
```diff
@@ -0,0 +1,37 @@
...
+RSpec.describe Glider, type: :model do
+  fixtures :all
+
+  describe '#name' do
+    context 'glider name' do # how to improve this?
+      subject(:glider) do
+        Glider.new(glider_type: 'Wave 2', size: 'S', manufacturer: 'Skyjam')
+      end
+
+      it 'returns "Wave 2 | Skyjam"' do # should I use '' and "" the other way around?
```
Comment (nerdinand) [inline spec/models/glider_spec.rb:12]: It's up to you. Really, the text for the `it` method call is just for documentation of what the test is supposed to test.

File: spec/models/glider_spec.rb [hunk 5]
```diff
@@ -0,0 +1,37 @@
+require 'rails_helper'
+
+RSpec.describe Glider, type: :model do
+  fixtures :all
+
+  describe '#name' do
+    context 'glider name' do # how to improve this?
```
Comment (nerdinand) [inline spec/models/glider_spec.rb:7]: If you only have one test case, such as here, you can leave the `context` off completely. If you have multiple text cases, then the contexts should describe what the difference in the test setup is between the cases.

File: spec/models/glider_spec.rb [hunk 6]
```diff
@@ -0,0 +1,37 @@
...
+      subject(:glider) do
+        Glider.new(glider_type: 'Wave 2', size: 'S', manufacturer: 'Skyjam')
+      end
+
+      it 'returns "Wave 2 | Skyjam"' do # should I use '' and "" the other way around?
+        expect(glider.name).to eq('Wave 2 S | Skyjam')
+      end # should I test also something else?
```
Comment (nerdinand) [inline spec/models/glider_spec.rb:14]: When deciding what cases to test you should think about the common ("happy") cases (such as this one) and about edge cases. 

For common cases, usually one or 2 tests are enough.

Edge cases are cases where inputs are special (empty string, nil, 0, negative numbers). Think about if any of the input to your test can realistically be one of these and how your code should react to them. For example in this method, can `glider_type`, `size` or `manufacturer` be `nil` or empty string? If yes, you should write some tests that cover these cases.

File: spec/models/glider_spec.rb [hunk 7]
```diff
@@ -0,0 +1,37 @@
...
+        Glider.new(color: 'black')
+      end
+
+      it 'is true' do
+        expect(glider.text_color?).to be true
+      end
+    end
+
+    context 'glider does not have color' do
+      subject(:glider) { Glider.new } # which style is better?
```
Comment (nerdinand) [inline spec/models/glider_spec.rb:30]: According to the [Ruby Style Guide](https://github.com/rubocop-hq/ruby-style-guide#single-line-blocks), you should use the `{}` block literal if your block is only one line long. However, if your line gets too long by doing this, prefer `do ... end`. For example, this would seem a bit long to me:
```
      subject(:glider) { Glider.new(glider_type: 'Wave 2', size: 'S', manufacturer: 'Skyjam') }
```

**Comment by nerdinand** [comment]
@hharen i created a PR for this so I can answer easily to your questions ;P

File: spec/models/day_spec.rb (hunk 1 above)
Comment (hharen) [reply spec/models/day_spec.rb:19]: Yes, I read that answer before asking but as you wrote, it seems author prefer let you subject. We also don’t call methods implicitly on subject so it would look like we don’t really need it.

Comment (hharen) [reply spec/models/day_spec.rb:19]: ...like Stack Overflow answer is very elaborate bug at the end it leaves you a bit confused...

File: spec/models/day_spec.rb (hunk 2 above)
Comment (hharen) [reply spec/models/day_spec.rb:36]: Yep, but I probably wouldn’t do it anyway, cause it’s less clear

File: spec/models/day_spec.rb (hunk 3 above)
Comment (hharen) [reply spec/models/day_spec.rb:81]: Clear

File: spec/models/glider_spec.rb (hunk 5 above)
Comment (hharen) [reply spec/models/glider_spec.rb:7]: 👍

File: spec/models/glider_spec.rb (hunk 6 above)
Comment (hharen) [reply spec/models/glider_spec.rb:14]: So I should check if nil returns nil or empty string? Hmmm

File: spec/models/glider_spec.rb (hunk 7 above)
Comment (hharen) [reply spec/models/glider_spec.rb:30]: 👍

File: spec/models/day_spec.rb (hunk 1 above)
Comment (nerdinand) [reply spec/models/day_spec.rb:19]: Yeah, i think it doesn't really matter. Personally, I like that this solution

```
subject(:day) do
  Day.new(students: [student1])
end

it 'returns false' do
  expect(day.has_student?(student2).to be false
end
```

shows you (without looking at the test itself) what the subject of the test is. And at the same time, inside the test, you don't have `subject`, but `day`, which tells you a bit more...

File: spec/models/glider_spec.rb (hunk 6 above)
Comment (nerdinand) [reply spec/models/glider_spec.rb]: So looking at the gliders migration, you only have `null: false` for `glider_type`, meaning that only `glider_type` cannot be nil. So you should do this:

Look at all the attributes and think about what `name` should return if they're `nil` or empty string. If it makes no sense that they can be `nil` or empty string, then don't handle that case in the test, but make it impossible (by implementing some validations).

See how writing tests makes you think about these valuable things you might otherwise not ask yourself? ;)

File: app/models/glider.rb [hunk 8]
```diff
@@ -13,7 +13,11 @@ class Glider < ApplicationRecord
   has_many :attendances
 
   def name
-    "#{glider_type} #{size} | #{manufacturer}"
+    if manufacturer.nil?
```
Comment (nerdinand) [inline app/models/glider.rb:16]: If `manufacturer` can also be empty string (I assume it could be), then the safer variant for this if statement would be `if manufacturer.blank?`. `blank?` captures both `nil` and empty string.

Comment (hharen) [reply app/models/glider.rb:16]: cool, I'll change that

**Comment by nerdinand** [comment]
@hharen did you not want to merge this? It's been open for forever...

**Comment by hharen** [comment]
> @hharen did you not want to merge this? It's been open for forever...

I wanted to add some more tests but I guess it makes sense to merge it now and add more later. You're right.

File: spec/models/day_spec.rb (hunk 1 above)
Comment (hharen) [reply spec/models/day_spec.rb:19]: Yeah, I like this solution. Thanks.
